service/backend/pchain/block.go

```diff
@@ -0,0 +1,445 @@
+	}
+
+	utxoBytes, err := b.pClient.GetRewardUTXOs(ctx, &api.GetTxArgs{
+		TxID:     txID,
+		Encoding: formatting.Hex,
+	})
+	if err != nil {
+		return err
+	}
+
+	utxos := []*avax.UTXO{}
+	for _, bytes := range utxoBytes {
+		utxo := avax.UTXO{}
+		_, err = b.codec.Unmarshal(bytes, &utxo)
+		if err != nil {
+			return err
+		}
+		utxos = append(utxos, &utxo)
+	}
+	out <- &pmapper.DependencyTx{
+		Tx:          &tx,
+		RewardUTXOs: utxos,
+	}
+
+	return nil
+}
+
+func (b *Backend) newTxParser(
+	ctx context.Context,
+	networkIdentifier *types.NetworkIdentifier,
+	dependencyTxs map[string]*pmapper.DependencyTx,
+) (*pmapper.TxParser, error) {
+	hrp, err := mapper.GetHRP(networkIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	chainIDs, err := b.getChainIDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	inputAddresses, err := pmapper.GetAccountsFromUTXOs(hrp, dependencyTxs)
+	if err != nil {
+		return nil, err
+	}
+
+	return pmapper.NewTxParser(false, hrp, chainIDs, inputAddresses, dependencyTxs), nil
+}
+
+func (b *Backend) getChainIDs(ctx context.Context) (map[string]string, error) {
+	if b.chainIDs == nil {
+		b.chainIDs = map[string]string{
+			ids.Empty.String(): mapper.PChainNetworkIdentifier,
+		}
+
+		cChainID, err := b.pClient.GetBlockchainID(ctx, mapper.CChainNetworkIdentifier)
+		if err != nil {
+			return nil, err
+		}
+		b.chainIDs[cChainID.String()] = mapper.CChainNetworkIdentifier
+
+		xChainID, err := b.pClient.GetBlockchainID(ctx, mapper.XChainNetworkIdentifier)
+		if err != nil {
+			return nil, err
+		}
+		b.chainIDs[xChainID.String()] = mapper.XChainNetworkIdentifier
+	}
+
+	return b.chainIDs, nil
+}
+
+func (b *Backend) getBlockDetails(ctx context.Context, index int64, hash string) (*indexer.ParsedBlock, error) {
+	if index <= 0 && hash == "" {
+		return nil, errMissingBlockIndexHash
+	}
+
+	blockHeight := uint64(index)
+
+	// Extract block id from hash parameter if it is non-empty, or from index if stated
+	if hash != "" {
+		height, err := b.getBlockHeight(ctx, hash)
+		if err != nil {
+			return nil, err
+		}
+
+		if blockHeight > 0 && height != blockHeight {
+			return nil, errMismatchedHeight
+		}
+		blockHeight = height
+	}
+
+	return b.indexerParser.ParseBlockAtIndex(ctx, blockHeight)
+}
+
+func (b *Backend) getBlockHeight(ctx context.Context, hash string) (uint64, error) {
+	blockID, err := ids.FromString(hash)
+	if err != nil {
+		return 0, err
+	}
+
+	blockBytes, err := b.pClient.GetBlock(ctx, blockID)
+	if err != nil {
+		return 0, err
+	}
+
+	var block blocks.Block
+	_, err = b.codec.Unmarshal(blockBytes, &block)
+	if err != nil {
+		return 0, err
+	}
+
+	return block.Height(), nil
+}
+
+func (b *Backend) isGenesisBlockRequest(ctx context.Context, index int64, hash string) (bool, error) {
+	genesisBlock, err := b.getGenesisBlock(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	if hash != "" {
+		return hash == genesisBlock.BlockID.String(), nil
+	}
+
+	return index == int64(genesisBlock.Height), nil
```

Comment:
The checks for hash and index are separated because in block or transaction lookup requests, one can either pass index or hash. If hash is present, we compare hash with genesis hash. If it is omitted, we then check the height. 

This line would default to true if a non-genesis block is looked up by hash only. I will add comments to clarify this.